### ISSUE TITLE
Make twisted.web.error compatible with Python 3

### ISSUE THREAD

**Issue by twisted-trac** [twisted/twisted#7845]
|[<img alt="dustin's avatar" src="https://avatars.githubusercontent.com/u/28673?s=50" width="50" height="50">](https://github.com/dustin)| @dustin reported|
|-|-|
|Trac ID|trac#7845|
|Type|enhancement|
|Created|2015-04-13 23:54:03Z|
|Branch|https://github.com/twisted/twisted/tree/py3-twisted-web-error-7845-2|

I thought this would be easy, but it has a nefarious dependency on twisted.web.template which required a lot of almost modules.

Attachments:

* [py3-twisted-web-error.patch](https://raw.githubusercontent.com/twisted/twistedmatrix.com-trac-attachments/trunk/ticket/f00/f00f74bd5eb6b0b2c79f8cb35b1545418c9d8610/6875dbe92b56982d32d16eb07f368a65f1023814.patch) (27193 bytes) - added by dustin on 2015-04-13 23:54:43Z - py3-twisted-web-error.patch

<details><summary>Searchable metadata</summary>

```
trac-id__7845 7845
type__enhancement enhancement
reporter__dustin dustin
priority__normal normal
milestone__ 
branch__branches_py3_twisted_web_error_7845_2 branches/py3-twisted-web-error-7845-2
branch_author__hawkowl__dustin__glyph hawkowl, dustin, glyph
status__closed closed
resolution__fixed fixed
component__core core
keywords__ 
time__1428969243814161 1428969243814161
changetime__1432277963137043 1432277963137043
version__None None
owner__hawkowl hawkowl

```
</details>

**Comment by twisted-trac** [comment]
|[<img alt="adiroiban's avatar" src="https://avatars.githubusercontent.com/u/204609?s=50" width="50" height="50">](https://github.com/adiroiban)<a name="note_1"></a>|@adiroiban commented|
|-|-|

(In [44457]) Branching to twisted.web.error-py3-7845.

**Comment by twisted-trac** [comment]
|[<img alt="adiroiban's avatar" src="https://avatars.githubusercontent.com/u/204609?s=50" width="50" height="50">](https://github.com/adiroiban)<a name="note_2"></a>|@adiroiban set owner to @dustin|
|-|-|

I tried to apply the patch on trunk but it failed

```
$ patch -p1 < ~/Downloads/py3-twisted-web-error.patch 
patching file twisted/python/compat.py
Hunk #1 FAILED at 11.
Hunk #2 succeeded at 384 (offset 81 lines).
1 out of 2 hunks FAILED -- saving rejects to file twisted/python/compat.py.rej
patching file twisted/python/dist3.py
Hunk #1 FAILED at 216.
Hunk #2 succeeded at 277 (offset 32 lines).
1 out of 2 hunks FAILED -- saving rejects to file twisted/python/dist3.py.rej
patching file twisted/python/urlpath.py
patching file twisted/web/_element.py
patching file twisted/web/_flatten.py
patching file twisted/web/_responses.py
patching file twisted/web/error.py
patching file twisted/web/template.py
patching file twisted/web/test/test_error.py
patching file twisted/web/test/test_web__responses.py

```

Can you please take a look at the patch.


I am not a big fan of diffs attached to Trac so I would be happy to get the diff from a Git branch hosted on GitHub and create a branch from that.

-----

I see that the new str2response method is only used inside the twisted/web/error.py module. Do we need it in twisted/web/_responses.py ?
Maybe just make it a private fuction in twisted/web/error.py  and skip the test for it.. but do keep the tests for the public errors using this code.

All tests need docstring... including the new one for str2response but I hope we can make that code private.

Also you will need to follow the Twisted coding convention, sorry. 2 new lines between methods ... 3 new lines between classed. I know that this is not easy.

Once patch applies on trunk I  can send the patch to the twistedchecker builder and we can see all errors.

Thanks!

**Comment by twisted-trac** [comment]
|[<img alt="dustin's avatar" src="https://avatars.githubusercontent.com/u/28673?s=50" width="50" height="50">](https://github.com/dustin)<a name="note_3"></a>|@dustin commented|
|-|-|

Here's the patch -- https://github.com/djmitche/Twisted/commit/8e537a20774e40d5e321a4afd5043c0295fb4f52.  I think it doesn't apply since your and hawkowl's py3 patches have landed since I wrote it.

I put str2response in _responses.py since it's about responses and is essentially the inverse of RESPONSES.  I'd ratehr keep it there, but I can add the unit test docstring.  I'll hold off on that until you've had the rest of a look.

**Comment by twisted-trac** [comment]
|[<img alt="adiroiban's avatar" src="https://avatars.githubusercontent.com/u/204609?s=50" width="50" height="50">](https://github.com/adiroiban)<a name="note_4"></a>|@adiroiban commented|
|-|-|

Thanks. Let's put this ticket back in the review queue.

**Comment by twisted-trac** [comment]
|[<img alt="adiroiban's avatar" src="https://avatars.githubusercontent.com/u/204609?s=50" width="50" height="50">](https://github.com/adiroiban)|@adiroiban removed owner|
|-|-|

**Comment by twisted-trac** [comment]
|[<img alt="adiroiban's avatar" src="https://avatars.githubusercontent.com/u/204609?s=50" width="50" height="50">](https://github.com/adiroiban)<a name="note_7"></a>|@adiroiban commented|
|-|-|

(In [44569]) Branching to twisted.web.error-py3-7845-2.

**Comment by twisted-trac** [comment]
|[<img alt="adiroiban's avatar" src="https://avatars.githubusercontent.com/u/204609?s=50" width="50" height="50">](https://github.com/adiroiban)<a name="note_8"></a>|@adiroiban set owner to @dustin|
|-|-|

Thanks for the git commit... Much easier to work with.

I solved the conflicts but please check that the merge is valid in the new branch.

-------

I don't want to force you to move str2response ... but I think that it should be renamed codeToPhrase or something to hint about `rfc2616 - 6.1.1 Status Code and Reason Phrase` and to not use leetspeak.

I am not sure what is Twisted policy regarding leetspeak in method names... so this is only a personal request.

Also, in case you want to have str2response in _response.py please update the module docstring: `HTTP response code definitions.` to something like `HTTP response code definitions and helpers to handle status code and reason phrases.`

--------

Changes looks good otherwise.

Many thanks for the new tests!

I think that instead of assertIdentical we can use assertIs as defined in stdlib unit test.

Please update docstrings for str2response testes and for the other new methods.

Docstring for new test methods are great. Thanks!

Also for class docstrings for RedirectWithNoLocationTests ... I am not sure if the class docstring should contain ' ... are initialized and displayed.' as this info should go into the test method docstring... and later we can extend RedirectWithNoLocationTests with tests which does not check for initialization and display.

Also, I think that "Tests for how" part can be left out as in the context of test code it is redundant.

In the case of RedirectWithNoLocationTests ... I think that any docstring is redundant :( ... I will open a discussion in twisted mail list and see what can be done.

For things like `self.assertEqual(e.message, "NO LOCATION to https://example.com")` can you please add explicit markup for text or  bytes ?


----------

I have sent the branch to buildbot and waiting for results.

I think that the branch is is good shape and it only need a bit of cleanup.

Please check the above comments and buildbot / twistedchecker results.

In case you don't have time to do the change please let us know and I will try to continue working on this.

Many thanks again!

**Comment by twisted-trac** [comment]
|[<img alt="adiroiban's avatar" src="https://avatars.githubusercontent.com/u/204609?s=50" width="50" height="50">](https://github.com/adiroiban)<a name="note_9"></a>|@adiroiban commented|
|-|-|

The following tests are failing on buildbots and only on my local system ...
they fain on py2.7 and py2.6

```
[FAIL]
Traceback (most recent call last):
  File "/home/buildslave/run/debian6-x86_64-py2.6-glib/Twisted/twisted/web/test/test_flatten.py", line 448, in test_string
    "Exception while flattening:\n"
  File "/home/buildslave/run/debian6-x86_64-py2.6-glib/Twisted/twisted/trial/_synctest.py", line 437, in assertEqual
    super(_Assertions, self).assertEqual(first, second, msg)
  File "/usr/lib/python2.6/unittest.py", line 350, in failUnlessEqual
    (msg or '%r != %r' % (first, second))
twisted.trial.unittest.FailTest: "Exception while flattening:\n  '01234567890123456789<...>1234567890123456789'\nRuntimeError: reason\n" != "Exception while flattening:\n  '01234567890123456789<...>01234567890123456789'\nRuntimeError: reason\n"

twisted.web.test.test_flatten.FlattenerErrorTests.test_string
===============================================================================
[FAIL]
Traceback (most recent call last):
  File "/home/buildslave/run/debian6-x86_64-py2.6-glib/Twisted/twisted/web/test/test_flatten.py", line 469, in test_unicode
    "Exception while flattening:\n"
  File "/home/buildslave/run/debian6-x86_64-py2.6-glib/Twisted/twisted/trial/_synctest.py", line 437, in assertEqual
    super(_Assertions, self).assertEqual(first, second, msg)
  File "/usr/lib/python2.6/unittest.py", line 350, in failUnlessEqual
    (msg or '%r != %r' % (first, second))
twisted.trial.unittest.FailTest: "Exception while flattening:\n  u'01234567\\u2603901234567\\u2603901234567\\u2603901234567\\u2603901234567\\u2603901234567\\u2603901234567\\u2603901234567\\u2603901234567\\u2603901234567\\u26039'\nRuntimeError: reason\n" != "Exception while flattening:\n  u'01234567\\u2603901234567\\u26039<...>01234567\\u2603901234567\\u26039'\nRuntimeError: reason\n"

twisted.web.test.test_flatten.FlattenerErrorTests.test_unicode
```

**Comment by twisted-trac** [comment]
|[<img alt="dustin's avatar" src="https://avatars.githubusercontent.com/u/28673?s=50" width="50" height="50">](https://github.com/dustin)<a name="note_10"></a>|@dustin commented|
|-|-|

I don't much care for the trivialities of naming or docstrings.  As you'd prefer I'll call it `codeToPhrase` and put it in `error.py` as a private function.  The docstrings for the new tests classes are copied from the first test in the file, and I agree that there's not much to say there -- the methods barely have behavior, not to speak of the classes.  But `RedirectWithNoLocationTests`'s docstring is definitely wrong.

I'll work on the test errors.

**Comment by twisted-trac** [comment]
|[<img alt="adiroiban's avatar" src="https://avatars.githubusercontent.com/u/204609?s=50" width="50" height="50">](https://github.com/adiroiban)<a name="note_11"></a>|@adiroiban commented|
|-|-|

In case my comments don't make sense you should not follow them.

Thanks for the feedback!

**Comment by twisted-trac** [comment]
|[<img alt="dustin's avatar" src="https://avatars.githubusercontent.com/u/28673?s=50" width="50" height="50">](https://github.com/dustin)<a name="note_12"></a>|@dustin removed owner|
|-|-|

They make sense - I'm just not going to argue :)

Oh, and as for the encoding of the message - it's always a native string: bytes (`str`) on python 2 and unicode (`str`) on python 3.

I've pushed a new version to the branch.  Aside from what was requested, I moved two of the `FlattenerError` tests out of `_flatten.py` since they were testing the class's behavior, not the flattener.  I adjusted how that error is rendered, too, to match the existing behavior regarding exactly where the ellipsis is inserted.  I also adjusted the test to expect a `..u'..` within the exception on Python 2, and not on Python 3.  This now passes both under Python 2 and 3 for me.

**Comment by twisted-trac** [comment]
|[<img alt="adiroiban's avatar" src="https://avatars.githubusercontent.com/u/204609?s=50" width="50" height="50">](https://github.com/adiroiban)<a name="note_13"></a>|@adiroiban commented|
|-|-|

I don't know what happened with the branch attached to this ticket...

The new branch should be https://github.com/twisted/twisted/compare/trunk...twisted.web.error-py3-7845-2

I could not find the branch you were referring the previous comment.

Can you please add an explicit link to it?

Thanks!

**Comment by twisted-trac** [comment]
|[<img alt="dustin's avatar" src="https://avatars.githubusercontent.com/u/28673?s=50" width="50" height="50">](https://github.com/dustin)<a name="note_14"></a>|@dustin commented|
|-|-|

https://github.com/twisted/twisted/compare/trunk...djmitche:py3-twisted-web-error

**Comment by twisted-trac** [comment]
|[<img alt="glyph's avatar" src="https://avatars.githubusercontent.com/u/716529?s=50" width="50" height="50">](https://github.com/glyph)<a name="note_15"></a>|@glyph commented|
|-|-|

(In [44706]) Branching to py3-twisted-web-error-7845.

**Comment by twisted-trac** [comment]
|[<img alt="glyph's avatar" src="https://avatars.githubusercontent.com/u/716529?s=50" width="50" height="50">](https://github.com/glyph)<a name="note_16"></a>|@glyph commented|
|-|-|

I put the branch into SVN to give the builders a crack at it; initial results are promising; one pyflakes issue.

**Comment by twisted-trac** [comment]
|[<img alt="glyph's avatar" src="https://avatars.githubusercontent.com/u/716529?s=50" width="50" height="50">](https://github.com/glyph)<a name="note_17"></a>|@glyph commented|
|-|-|

There's already a separate ticket, [#7811](https://github.com/twisted/twisted/issues/7811), for porting `twisted.web.template`, and that work should probably be completed separately.  Sorry for the poor record of dependencies :-\.  (This comment doesn't constitute a full review, so I'm not removing the tag.)

**Comment by twisted-trac** [comment]
|[<img alt="dustin's avatar" src="https://avatars.githubusercontent.com/u/28673?s=50" width="50" height="50">](https://github.com/dustin)<a name="note_18"></a>|@dustin commented|
|-|-|

This doesn't port `twisted.web.template` - just moves it to the "almost" set since it's imported (at runtime) by `twisted.web.error`

**Comment by twisted-trac** [comment]
|[<img alt="glyph's avatar" src="https://avatars.githubusercontent.com/u/716529?s=50" width="50" height="50">](https://github.com/glyph)<a name="note_19"></a>|@glyph commented|
|-|-|

Oh.  Well, given that [#7811](https://github.com/twisted/twisted/issues/7811) doesn't actually have any code on it perhaps we should just proceed then :).

**Comment by twisted-trac** [comment]
|[<img alt="dustin's avatar" src="https://avatars.githubusercontent.com/u/28673?s=50" width="50" height="50">](https://github.com/dustin)<a name="note_20"></a>|@dustin commented|
|-|-|

Proceed with the review?  I neither ported nor intend to port `twisted.web.template`, so I'm not sure how [#7811](https://github.com/twisted/twisted/issues/7811) is relevant at all.

**Comment by twisted-trac** [comment]
|[<img alt="glyph's avatar" src="https://avatars.githubusercontent.com/u/716529?s=50" width="50" height="50">](https://github.com/glyph)<a name="note_21"></a>|@glyph commented|
|-|-|

Replying to [dustin](#note_20):
> Proceed with the review?  I neither ported nor intend to port `twisted.web.template`, so I'm not sure how [#7811](https://github.com/twisted/twisted/issues/7811) is relevant at all.

It's not relevant. What I was saying was, please disregard my earlier comment, since there are no conflicts to worry about or inverted dependencies, this should be reviewed (and hopefully landed) as-is.

**Comment by twisted-trac** [comment]
|[<img alt="hawkowl's avatar" src="https://avatars.githubusercontent.com/u/3307100?s=50" width="50" height="50">](https://github.com/hawkowl)<a name="note_22"></a>|@hawkowl commented|
|-|-|

(In [44821]) add error as a module and test_error as the tests (refs [#7845](https://github.com/twisted/twisted/issues/7845))

**Comment by twisted-trac** [comment]
|[<img alt="hawkowl's avatar" src="https://avatars.githubusercontent.com/u/3307100?s=50" width="50" height="50">](https://github.com/hawkowl)<a name="note_23"></a>|@hawkowl commented|
|-|-|

(In [44822]) add urlparse to __all__ (refs [#7845](https://github.com/twisted/twisted/issues/7845))

**Comment by twisted-trac** [comment]
|[<img alt="hawkowl's avatar" src="https://avatars.githubusercontent.com/u/3307100?s=50" width="50" height="50">](https://github.com/hawkowl)<a name="note_24"></a>|@hawkowl commented|
|-|-|

Rather than send it through another cycle, I just went through and fixed some minor things (pyflakes, twistedchecker that the bots picked up on, some ```__future__}} module imports).

As a reviewer, I am happy with this. But as I *have* made some changes, I'm putting this up for re-review. Builders spun.

Thanks for the patch, dustin :```{__future__}} module imports).

As a reviewer, I am happy with this. But as I *have* made some changes, I'm putting this up for re-review. Builders spun.

Thanks for the patch, dustin :)

**Comment by twisted-trac** [comment]
|[<img alt="hawkowl's avatar" src="https://avatars.githubusercontent.com/u/3307100?s=50" width="50" height="50">](https://github.com/hawkowl)<a name="note_25"></a>|@hawkowl commented|
|-|-|

As I said, putting it back up for review.

**Comment by twisted-trac** [comment]
|[<img alt="adiroiban's avatar" src="https://avatars.githubusercontent.com/u/204609?s=50" width="50" height="50">](https://github.com/adiroiban)<a name="note_26"></a>|@adiroiban set owner to @hawkowl|
|-|-|

Changes look good to merge.

Do we need CodeToMessageTests ? I was expecting to keep it private and only have test for the public code using it.

In this way we can refactor the code (including a version which does not use this method) without having to care about CodeToMessageTests

Hawkie, I am putting this on your back to review my comment and merge :)

Thank you all!

**Comment by twisted-trac** [comment]
|[<img alt="hawkowl's avatar" src="https://avatars.githubusercontent.com/u/3307100?s=50" width="50" height="50">](https://github.com/hawkowl)<a name="note_27"></a>|@hawkowl commented|
|-|-|

Hi adi,

Direct unit tests are always better than having it tested by non-direct integration tests -- if we refactor it, we'll just remove the private function and the tests.

Will merge.

**Comment by twisted-trac** [comment]
|[<img alt="hawkowl's avatar" src="https://avatars.githubusercontent.com/u/3307100?s=50" width="50" height="50">](https://github.com/hawkowl)<a name="note_28"></a>|@hawkowl commented|
|-|-|

(In [44907]) Branching to py3-twisted-web-error-7845-2.

**Comment by twisted-trac** [comment]
|[<img alt="hawkowl's avatar" src="https://avatars.githubusercontent.com/u/3307100?s=50" width="50" height="50">](https://github.com/hawkowl)<a name="note_29"></a>|@hawkowl set status to `closed`|
|-|-|

(In [44911]) Merge py3-twisted-web-error-7845-2: Port twisted.web.error to Python 3

Author: dustin
Reviewers: adiroiban, hawkowl
Fixes: [#7845](https://github.com/twisted/twisted/issues/7845)
